src/backend/distributed/relay/relay_event_utility.c

```diff
@@ -613,11 +613,17 @@ AppendShardIdToName(char **name, uint64 shardId)
 	shardIdAndSeparatorLength = strlen(shardIdAndSeparator);
 
 	/*
-	 * If *name strlen is < (NAMEDATALEN - shardIdAndSeparatorLength),
-	 * it is safe merely to append the separator and shardId.
+	 * If *name strlen is < (NAMEDATALEN - shardIdAndSeparatorLength - 1),
+	 * it is safe merely to append the separator and shardId. We truncate
+	 * down to 63 instead of 64 characters because postgres generates an
+	 * array type named _<relation name truncated to 63 characters> during
```

Comment:
`makeArrayTypeName` appears to be the function that actually does this, by the way. Maybe add a reference to that…

It uses a function called `truncate_identifier` that might be useful to clean up the multi-byte truncation code we added, since it has an ominous comment about _avoid[ing] using `%.*s` here because it can misbehave if the data is not valid in what libc thinks is the prevailing encoding_.